event_routing_backends/processors/xapi/event_transformers/exam_events.py

```diff
@@ -0,0 +1,110 @@
+"""
+Transformers for enrollment related events.
+"""
+
+from tincan import Activity, ActivityDefinition, Extensions, LanguageMap, Verb
+
+from event_routing_backends.processors.xapi import constants
+from event_routing_backends.processors.xapi.registry import XApiTransformersRegistry
+from event_routing_backends.processors.xapi.transformer import XApiTransformer
+
+
+class BaseExamTransformer(XApiTransformer):
+    """
+    Base transformer for exam events.
+    """
+
+    def get_object(self):
+        """
+        Get object for xAPI transformed event.
+
+        Returns:
+            `Activity`
+        """
+        object_id = self.get_data("event.exam_content_id")
+
+        return Activity(
+            id=object_id,
+            definition=ActivityDefinition(
+                type=constants.XAPI_ACTIVITY_ASSESSMENT,
+                name=LanguageMap(**({constants.EN: self.get_data("event.exam_name")})),
+                extensions=Extensions(
+                    {
+                        constants.XAPI_ACTIVITY_PRACTICE_EXAM: self.get_data(
+                            "event.exam_is_practice_exam"
+                        ),
+                        constants.XAPI_ACTIVITY_PROCTORED_EXAM: self.get_data(
+                            "event.exam_is_proctored"
+                        ),
+                        constants.XAPI_ACTIVITY_COURSE: self.get_data("event.course_id"),
+                    }
+                ),
+            ),
+        )
+
+    def get_context_activities(self):
+        context_activities = super().get_context_activities()
+
+        context_activities.grouping = [
+            Activity(
+                id=self.get_data("event.attempt_code"),
+                definition=ActivityDefinition(
+                    type=constants.XAPI_ACTIVITY_EXAM_ATTEMPT,
+                    name=LanguageMap({constants.EN: self.get_data("event.exam_name")}),
+                    extensions=Extensions(
+                        {
+                            constants.XAPI_CONTEXT_ATTEMPT_STARTED: self.get_data(
+                                "event.attempt_started_at"
+                            ),
+                            constants.XAPI_CONTEXT_ATTEMPT_COMPLETED: self.get_data(
+                                "event.attempt_completed_at"
+                            ),
+                            constants.XAPI_CONTEXT_ATTEMPT_TIME: self.get_data(
+                                "event.attempt_event_elapsed_time_secs"
+                            ),
+                            constants.XAPI_CONTEXT_ID: self.get_data("event.attempt_id"),
+                        }
+                    ),
+                ),
+            ),
+        ]
+
+        return context_activities
+
+
+@XApiTransformersRegistry.register("edx.special_exam.timed.attempt.created")
```

Comment:
Unsure but maybe we could use the same transformer for proctored and practice exam attempts.

Reply:
I think that would be ok as long as we don't need more than like 1 "if" statement in there to state the verb or activity type, otherwise it gets convoluted and it's easier to break them up.

Reply:
Something like this to extract the type
```python
self.event['name'].split(".")[2]
```

Reply:
Done